### PR TITLE
[Deps] Update to Clang-16

### PR DESCRIPTION
**Pull request by yijiem**

### CODE REVIEW

**Comment by yijiem** [comment]
Manual invocation of the `grpc/core/master/linux/grpc_portability` job on this PR seems to succeed: https://fusion2.corp.google.com/invocations/c4fec396-5477-434f-9ae9-896f445f10eb/log

File: src/core/ext/filters/client_channel/lb_policy/ring_hash/ring_hash.cc [hunk 1]
```diff
@@ -37,6 +37,8 @@
 #include "absl/strings/string_view.h"
 #include "absl/types/optional.h"
 
+#include "src/core/lib/iomgr/resolved_address.h"
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/lb_policy/ring_hash/ring_hash.cc:40]: This header reordering change looks incorrect.  I think clang-format is being confused by the `#define XXH_INLINE_ALL` on line 42.

**Review by markdroth — COMMENTED** [review]
One quick drive-by comment.

File: .clang-tidy [hunk 2]
```diff
@@ -1,15 +1,27 @@
...
 #   Some macros need to be defined for portability purpose; e.g. _BSD_SOURCE.
 #
 # - modernize-redundant-void-arg
 #   Some source should be strictly C99 and func(void) should be used.
 #
+# - google-readability-casting
+#   https://github.com/llvm/llvm-project/issues/57959
+#
+# - readability-redundant-string-cstr
+#   This is good to have, but needs a large refactor across the codebase.
```
Comment (veblush) [inline .clang-tidy:23]: Please move this comment down below in the secion for future work.

**Review by veblush — APPROVED** [review]

File: .clang-tidy (hunk 2 above)
Comment (yijiem) [reply .clang-tidy]: Done.

File: src/core/ext/filters/client_channel/lb_policy/ring_hash/ring_hash.cc (hunk 1 above)
Comment (yijiem) [reply src/core/ext/filters/client_channel/lb_policy/ring_hash/ring_hash.cc]: Thanks! Reverted.

**Pull request merged**